### ISSUE TITLE
Issues with running the  Oracle Tracker

### ISSUE THREAD

**Issue by sgk98** [phil-bergmann/tracking_wo_bnw#75]
Hello,
I was trying to reproduce the experiments with the Oracle Tracker and I ran into a few issues:

1) When running the script ```python experiments/scripts/test_tracktor.py with oracle```, I get the following error in line 357 of ```oracle_tracker.py```:
```self.obj_detect.load_image(blob['data'][0], blob['im_info'][0]) KeyError: 'data'```
Looking further into this issue, I noticed that blob is coming from the DataLoader and the keys defined for the blob are ```dict_keys(['img', 'dets', 'img_path', 'gt', 'vis'])```

This issue is not present while running the regular Tracktor script however.
Any suggestions to fix this issue would very most helpful.
Thanks!

**Comment by sgk98** [comment]
Some more additional details:
Till the commit ```30b89259243251c816e45ef70d13dd3df49b7b48```("Restructe siamese to reid and remove frcnn and fpn files") ```tracker.py``` was also using the keys ```data``` and ```im_info```.
However, in the next, ```tracker.py``` was updated while ```oracle_tracker.py``` thereby leading to this issue. I am unsure about how to fix ```oracle_tracker.py``` therefore it would be very useful if you could help out with this.

Thanks!

**Comment by timmeinhardt** [comment]
Hello, sorry for my late reply. I just updated the code and fixed the oracle trackers. I also ran some evaluations to check if the results are still analogous to the paper, i.e., the improvements of each oracle ranks similar to the numbers reported in the paper. Everything seemed reasonable. If you encounter any strange behaviour or numbers, do not hesitate to write here again.

**Comment by sgk98** [comment]
Thank you so much for your reply!
The updated code seems to work fine in most cases, however, when Oracle-kIll is not enabled, there is an error on line 441 of ```oracle_tracker.py```  saying ```nms() missing 1 required positional argument: 'iou_threshold'``` 
Looking at the other uses of the ```nms()`` function, it looks like there is a missing argument, however, I am not sure what variable needs to be passed.

**Comment by igygi** [comment]
Hi @sgk98, may I know how you ran `test_tracktor.py` with `oracle`?

**Comment by sgk98** [comment]
The command I used was ```python experiments/scripts/test_tracktor.py with oracle``` with the oracle options appropriately set in ```experiments/cfgs/scripts/oracle_tracktor.yaml```

**Comment by timmeinhardt** [comment]
> Thank you so much for your reply!
> The updated code seems to work fine in most cases, however, when Oracle-kIll is not enabled, there is an error on line 441 of `oracle_tracker.py` saying `nms() missing 1 required positional argument: 'iou_threshold'`
> Looking at the other uses of the `nms()` function, it looks like there is a missing argument, however, I am not sure what variable needs to be passed.
> 
> EDIT:
> A quick fix to this issue seems to be to replace line 441 with `keep = nms(self.get_pos(),person_scores ,self.regression_nms_thresh)` (I took this from `tracker.py`).

Yes, this is the appropriate fix. I will integrate this into the repository. Thank you!

> However, I am unsure if this is correct, but it seems to work. A confirmation from your side would be helpful.
> 
> A couple of other questions about the Oracle Tracker results:
> 
> 1. The paper lists the original Tracktor results(MOTA, IDF1)  with Faster-RCNN detections on MOT17 as (61.5,61.1) while on running the code, I get only (59.6,59.7). Even Tracktor++V2 gives only (61.7,65.2) (The IDF1 is significantly higher, but MOTA is still lesser than the original Tracktor++.) Is this behavior expected or am I supposed to change some hyperparameters? I am confused about this because in all my experiments so far(apart from the Oracle experiments), Tracktor++v2 has significantly outperformed the ICCV19 results.

Tracktor++v2 applies a new object detector model. Small changes in the MOTA and IDF1 are to be expected. This includes that the MOTA on the Faster RCNN detections alone is slightly worse than for the original ICCV19 version. I think you should not compare v2 vs. ICCV19 but use one of them and only check if the internal comparisons make sense, i.e., if the different oracle trackers improve on the correct aspect and the general order of performance improvements is consistent.

> 2. The results obtained by Oracle ReID(when applied alone) and the Siamese network trained for ReID seems to be giving the same performance(marginally better performance than Tracktorv2, but worse than Tracktor++v2). I'm guessing this could be some issue with the ReID implementation in the `oracle_tracker.py`?

Yes, this indeed seems be an error. Tracktor++v2 + Oracle reID should improve on Tracktor++v2 alone.

**Comment by sgk98** [comment]
Thank you for the update!
Yes, the only remaining issue with the Oracle trackers seems to be that Oracle ReID doesn't seem to be improving performance in any of the combinations that I've tried(Tracktor++v2 + OracleReID, Tracktor++V2 + MM+Kill+ReiD etc.). It would be great if you could fix this issue.

**Comment by drimpossible** [comment]
@timmeinhardt are there any updates on when the fix might be available?
